### DISCUSSION TITLE
Error occurs when receiving flutter firestore data

### DISCUSSION THREAD

**Discussion by hhamja** [AntlerVC/firetable#1259] [Support / Q&A] [answered]
Using flutter + rowy, I created a collection called product and put data into it. Then, I tried to get the entire product data, but the following error occurred.

FirebaseException ([cloud_firestore/permission-denied] The caller does not have permission to execute the specified operation.)

I googled it and found out that it is caused by firestore rules.
First of all, since my app does not force user sign-up, you should be able to get product data even when you are not logged in. But in rowy, can't I set it to read data in a non-login state?

In rowy product Access controls, 
All permissions of OWNER,EDITOR ,VIEWER, ADMIN  have been checked. 

--- -----------------------FireStore role, below ------------------------------
rules_version = '2';
service cloud.firestore {
	match /databases/{database}/documents {
		// ---------------------------------------------------------
		// Rowy rules start
		// ---------------------------------------------------------
		// Automatically generated and managed by Rowy
		// ⚠️ Do not edit this block directly, as it will be overwritten
		// To edit the security rules, please do it on different block or manage from rowy settings
		// initialized on 5/11/2023
		match /{collectionId}/{docId} {
			allow read: if colRule(["product","ingredient"], ["ADMIN","EDITOR","VIEWER","OWNER"]);
			allow create, update, delete: if colRule(["product"], ["ADMIN","EDITOR","VIEWER","OWNER"]);
			allow create, update, delete: if colRule(["ingredient"], ["ADMIN","OWNER"]);
			
			match /{subTableCollectionID}/{document=**}{
				allow read, create, update, delete: if colRule(["product"], ["ADMIN","EDITOR","VIEWER","OWNER"]);
				
			}
			function colRule(collections, roles){
				return collectionId in collections && hasAnyRole(roles);
			}
			
		}
		// ---------------------------------------------------------
		// Rowy rules end
		// ---------------------------------------------------------
		// Allow admins to read and write all documents
		match /{document=**} {
			allow read, write: if hasAnyRole(["ADMIN", "OWNER"]);
		}
		
		// Rowy: Allow signed in users to read Rowy configuration and admins to write
		match /_rowy_/{docId} {
			allow read: if request.auth.token.roles.size() > 0;
			allow write: if hasAnyRole(["ADMIN", "OWNER"]);
			match /{document=**} {
				allow read: if request.auth.token.roles.size() > 0;
				allow write: if hasAnyRole(["ADMIN", "OWNER"]);
			}
		}
		// Rowy: Allow users to edit their settings
		match /_rowy_/userManagement/users/{userId} {
			allow get, update, delete: if isDocOwner(userId);
			allow create: if request.auth.token.roles.size() > 0;
		}
		// Rowy: Allow public to read public Rowy configuration
		match /_rowy_/publicSettings {
			allow get: if true;
		}
		
		// Rowy: Utility functions
		function isDocOwner(docId) {
			return request.auth != null && (request.auth.uid == resource.id || request.auth.uid == docId);
		}
		function hasAnyRole(roles) {
			return request.auth != null && request.auth.token.roles.hasAny(roles);
		}
	}
}

----------------------------------------------------------------------------------------------------------------------------

It's very frustrating. It is mentioned not to touch it directly in the Firestore rules, but the setting in rowy for the case where user is null is not visible anywhere. 

Please help me

**Comment by jocypher** [comment]
make sure you have edited your rules for Both the storage and cloud_firestore(firestore database) to be able to read and write .
`service cloud.firestore {
3
  match /databases/{database}/documents {
4
    match /{document=**} {
5
      allow read, write;
6
    }
7
  }
8
}
`

**Comment by BeeBombshell** [comment, marked as answer]
Hey @CodeKinggggggg, the error message you're encountering indicates a permission issue with Firestore. In your Rowy configuration, you have defined access controls for the "product" collection, but it seems that the rules are not allowing read access to non-logged-in users.

To address this issue, you can modify your Firestore rules to allow read access to the "product" collection even when the user is not logged in. Here's an updated version of your Firestore rules with the necessary changes:

```js
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    // Allow read access to the "product" collection for non-logged-in users
    match /product/{document=**} {
      allow read: if true;
      allow create, update, delete: if colRule(["product"], ["ADMIN","EDITOR","VIEWER","OWNER"]);
    }

    // Rest of the existing rules...
  }
}
```
